overlord/snapstate/snapmgr.go

```diff
@@ -780,6 +781,26 @@ func (m *SnapManager) atSeed() error {
 	return nil
 }
 
+func (m *SnapManager) ensureDiskSpace() error {
+	// Require at least 100Mb; core16 snap takes 97Mb, so this should allow
+	// for core snap refresh. Newer cores are smaller.
+	// XXX: we could try to be smarter here and require a minimum size
+	// needed for particular core version?
+	const requiredSpace = 100 * 1024 * 1024
+	path := dirs.SnapdStateDir(dirs.GlobalRootDir)
+	err := osutilCheckFreeSpace(path, requiredSpace)
+	if _, ok := err.(*osutil.NotEnoughDiskSpaceError); ok {
+		m.state.Lock()
+		defer m.state.Unlock()
+		// TODO: clean download cache and re-try osutilCheckFreeSpace before
+		// reporting the error.
+		m.state.Warnf(i18n.G("the available disk space at %q is less than %s, this may prevent refreshes and proper functioning of snapd"),
```

Comment:
As a side note, the check for duplicate warnings is naive and we should keep in mind to not include anything that may vary between ensure runs (eg. free disk space), otherwise this could trigger multiple warnings

Reply:
Ah yes, great point, I forgot this can generate varying message on each ensure. Short term we could just issue a generic warning with constant message ("low disk"), then we could enhance warnings api with something like AddWarning(warningKey, message string).